File: easy/buddy_strings/buddy_strings.py
class Solution(object):
    def buddyStrings(self, A, B):
        """
        :type A: str
        :type B: str
        :rtype: bool
        """
        if len(A) != len(B):
            return False

        places_of_difference = []
        for i in range(0, len(A)):
            if A[i] != B[i]:
                places_of_difference.append(i)
            if len(places_of_difference) > 2:
                return False

        if len(places_of_difference) == 2:
            first, second = places_of_difference[0], places_of_difference[1]
            if A[first] != B[second] or A[second] != B[first]:
                return False

        if len(places_of_difference) == 1:
            return False

        if len(places_of_difference) == 0:
            for c in A:
                if A.count(c) >= 2:
                    return True
            return False

        return True


s = Solution()
print("Solution 1 : ", s.buddyStrings("ab", "ba"))
print("Solution 2 : ", s.buddyStrings("ab", "ab"))
print("Solution 3 : ", s.buddyStrings("aa", "aa"))
print("Solution 4 : ", s.buddyStrings("aaaaaaabc", "aaaaaaacb"))
print("Solution 5 : ", s.buddyStrings("", "aa"))

File: easy/array/buddy_strings/buddy_strings.go
package main

import (
	"fmt"
	"strings"
)

func buddyStrings(A string, B string) bool {
	if len(A) != len(B) {
		return false
	}

	placesOfDifference := make([]int, 0, 3)
	for i := 0; i < len(A); i++ {
		if A[i] != B[i] {
			placesOfDifference = append(placesOfDifference, i)
		}
		if len(placesOfDifference) > 2 {
			return false
		}
	}

	if len(placesOfDifference) == 2 {
		first, second := placesOfDifference[0], placesOfDifference[1]
		if A[first] != B[second] || A[second] != B[first] {
			return false
		}
	}

	if len(placesOfDifference) == 1 {
		return false
	}

	if len(placesOfDifference) == 0 {
		for i := 0; i < len(A); i++ {
			if strings.Count(A, string(A[i])) >= 2 {
				return true
			}
		}
		return false
	}

	return true
}

func main() {
	fmt.Println("Solution 1 : ", buddyStrings("ab", "ba"))
	fmt.Println("Solution 2 : ", buddyStrings("ab", "ab"))
	fmt.Println("Solution 3 : ", buddyStrings("aa", "aa"))
	fmt.Println("Solution 4 : ", buddyStrings("aaaaaaabc", "aaaaaaacb"))
	fmt.Println("Solution 5 : ", buddyStrings("", "aa"))
}
